src/realm/obj.cpp

```diff
@@ -1945,57 +2023,111 @@ void Obj::assign(const Obj& other)
             auto sz = src_list->size();
             dst_list->clear();
             for (size_t i = 0; i < sz; i++) {
-                Mixed val = src_list->get_any(i);
-                dst_list->insert_any(i, val);
+                Mixed value = src_list->get_any(i);
+                dst_list->insert_any(i, handle_object_to_copy(value));
+            }
+        }
+        else if (col.get_attrs().test(col_attr_Set)) {
+            auto src_set = other.get_setbase_ptr(col);
+            auto dst_set = get_setbase_ptr(col);
+            auto sz = src_set->size();
+            dst_set->clear();
+            for (size_t i = 0; i < sz; ++i) {
+                Mixed value = src_set->get_any(i);
+                dst_set->insert_any(handle_object_to_copy(value));
+            }
+        }
+        else if (col.get_attrs().test(col_attr_Dictionary)) {
+            auto src_dictionary = other.get_dictionary_ptr(col);
+            auto dst_dictionary = get_dictionary_ptr(col);
+            auto sz = src_dictionary->size();
+            dst_dictionary->clear();
+            for (size_t i = 0; i < sz; ++i) {
+                const auto& [key, value] = src_dictionary->get_pair(i);
+                TableRef target_table;
+                ObjKey target_obj_key;
+                if (is_outgoing_embedded_link(value, target_table, target_obj_key)) {
+                    auto obj = dst_dictionary->create_and_insert_linked_object(key);
+                    obj.copy(target_table->get_object(target_obj_key));
```

Comment:
Yes, it is tested by `SECTION("change table to embedded - violate embedded object constraints")`, 
and no, this is not what it is going to happen. 

  `dst_dictionary` is the copied dictionary (it belongs to the just cloned object, so at this stage it is effetely empty or in general cannot contain a duplicate key, unless the `src_dictionary` does), 

The invariant that holds at this stage is that the `dst_dictionary` belongs to the just cloned object, so it can never be doing something bad with the original data.